### PR TITLE
YOLO

### LINKED ISSUE #6341: [RFC] Support DarkNet and YOLO series models
### 🚀 The feature

YOLO aka. You Only Look Once, which is a vibrant series of object detection models since the release of Joseph Redmon [You Only Look Once: Unified, Real-Time Object Detection](https://arxiv.org/abs/1506.02640).

So far a couple of more notable implementations are as follows (all PyTorch):

- [ ] [YOLOv3](https://arxiv.org/abs/1804.02767) 2018. Cited by 14369 [^yolov3-ultralytics][^yolov3-mmdet]
- [ ] [YOLOv4](https://arxiv.org/abs/2004.10934) 2020. Cited by 4905
- [ ] YOLOv5 2020. Starred at GitHub 29.3k [^yolov5-ultralytics]
- [ ] [YOLOX](https://arxiv.org/abs/2107.08430) 2021. Cited by 299 [^yolox]
- [ ] [YOLOv7](https://arxiv.org/abs/2207.02696) 2022. [^yolov7]

### Motivation, pitch

Until now, one of the most successful ones is probably YOLOv5. YOLOv5 is great, and they have also built up a very friendly community. We don't intend to copy YOLOv5 into TorchVision, our main goal here is to make training SoTA models easier and share reusable subcomponents to build the next SoTA models in the same/proxy family.[^yolo-keras]

### The License

YOLO{v5/v7} are built under the [GPL-3.0 license](https://github.com/ultralytics/yolov5/blob/master/LICENSE), and YOLOX is built under the [Apache-2.0 license](https://github.com/Megvii-BaseDetection/YOLOX/blob/main/LICENSE), it's easier to implement yolox first, and then build YOLO{v5/v7} upon it.

### What we should do

I have previously rewritten the code used in the inference part of yolov5 according to the style and specification of torchvision, and I can relicense that part to [BSD-3-Clause license](https://github.com/pytorch/vision/blob/main/LICENSE). The amount of work involved in the model inference part is not much with the help of YOLOX base code.

Data augmentation and model trainer engine will be the core of what we will do here.

The data augmentation section is in the planning list https://github.com/pytorch/vision/issues/6224 , and we have already merged some augmentation methods like https://github.com/pytorch/vision/pull/5825 , I think it would help us to build the next SoTA models with a new primitives like classification models.[^classification-primitives]

As TorchVision adds more and more models, it may be time to abstract out a simple trainer engine for sharing reusable subcomponents. It might be more appropriate to open a new thread for necessity and specific steps about this part.

### Alternatives

_No response_

### Additional context

[^yolov3-ultralytics]: https://github.com/ultralytics/yolov3/tree/v9.1
[^yolov3-mmdet]: https://github.com/open-mmlab/mmdetection/tree/master/configs/yolo
[^yolov5-ultralytics]: https://github.com/ultralytics/yolov5
[^yolox]: https://github.com/Megvii-BaseDetection/YOLOX
[^yolov7]: https://github.com/WongKinYiu/yolov7
[^yolo-keras]: https://github.com/keras-team/keras-cv/issues/622#issuecomment-1198063712
[^classification-primitives]: https://pytorch.org/blog/how-to-train-state-of-the-art-models-using-torchvision-latest-primitives/

### PR DESCRIPTION
**Pull request by senarvi**
A generic YOLO implementation that supports the most important features of YOLOv3, YOLOv4, YOLOv5, YOLOv7, Scaled-YOLOv4, and YOLOX. It includes networks that have been written in PyTorch, but the user can also load a network from a Darknet configuration file. The features such as matching predictions to targets have been implemented in a modular way, so that they can easily be replaced, or reused in different models. Target class labels may be specified as a matrix of class probabilities, allowing multi-label classification. Includes unit tests and complete type hints for static type checking.

This code is contributed with the premission of my employer Groke Technologies.

Fixes #6341

### CODE REVIEW

**Comment by oke-aditya** [comment]
cc @Nicolas

**Comment by senarvi** [comment]
Seems like I was able to fix most of the unit tests. The biggest job was to get the TorchScript compilation working. In order to fix it, I had to make the code a bit uglier in some places. For example, the target matching classes cannot be subclasses, because the JIT compilation can't handle subclasses. Also, the iou function and the cross-entropy function cannot be passed as function objects, which makes the loss computation a bit awkward. I don't know if there would be some way to make function objects work.

File: test/test_models.py [hunk 1]
```diff
@@ -783,6 +793,61 @@ def check_out(out):
...
+        num_samples = 20
+        if num_elems <= num_samples:
+            return tensor
+
+        ith_index = num_elems // num_samples
+        return tensor[ith_index - 1 :: ith_index]
+
+    def compute_mean_std(tensor):
+        # can't compute mean of integral tensor
+        tensor = tensor.to(torch.double)
```
Comment (vadimkantorov) [inline test/test_models.py:819]: is there a reason for using `torch.double` instead of `torch.float32`?
also, just in case there exists `torch.std_mean` function

Comment (senarvi) [reply test/test_models.py:819]: Thanks for reviewing @vadimkantorov . This function used to be an inner function of `test_detection_model()` and I just copied it outside the function so that other functions can call it. So I don't know if there has been a reason to cast it to double instead of float32, but based on the above comment it seems like float32 would be fine. Should I change it to float32 and also switch to `std_mean()`?

Comment (vadimkantorov) [reply test/test_models.py:819]: I'm an not one of maintainers, let's wait for a more official review :)

Comment (senarvi) [reply test/test_models.py:819]: @vadimkantorov all right, thanks for pointing that out though.

**Comment by patches11** [comment]
I'm attempting to test these models by training them on the COCO dataset, however I am unable to get close to the results reported in the original papers, for example yolo v7 here:

https://github.com/WongKinYiu/yolov7?tab=readme-ov-file#performance

Wondering if I am doing something wrong in training these, or if you can provide any guidance on what you are doing to train them and the results you are seeing?

**Comment by senarvi** [comment]
Hi @patches11 . I haven't trained models on YOLO recently. Also, I don't have access to a compute cluster for training these models anymore. I did check earlier that I can use YOLOv4 weights, so the forward pass should be correct. But there are lots of details used in model training, like mosaic and copy-paste augmentation. I feel like all the details are not mentioned in the papers. I'm not even sure how exactly the models were tested (what data and resolution were used). I also found gradient clipping to be important, even though it's not used in the papers. Maybe @FateScript can comment what we're still missing?

**Comment by patches11** [comment]
@senarvi thanks for the details, I will take a look at implementing those augmentations.

It does definitely seem like we don't get all the details in the papers

**Comment by senarvi** [comment]
Yeah, it will be interesting if we can get all the augmentations and training tricks exactly as in the papers. That way we could get a fair comparison between YOLO and other architectures. I think YOLOv7 uses 1280x1280 input size, which consists of four 640x640 tiles. So for each network input you sample four images, which makes it a bit more complicated. For the copy-paste augmentation we need segmentation masks, in addition to the bounding boxes. I think during testing they sort the images so that for each batch you get as similar sizes as possible, so that you don't have to resize the images as much. Still, if I'm right, the test results vary a little bit, depending on the batch size.

**Comment by discort** [comment]
@senarvi  
Thanks for sharing your implementation. Any plans to finish this PR?

**Comment by senarvi** [comment]
@discort I was left waiting for someone to review this PR. I'm not sure if there's interest in merging this to torchvision... there hasn't been much activity in 1.5 years. More or less the same implementation was merged to lightning-bolts, though: https://github.com/Lightning-Universe/lightning-bolts/tree/master/src/pl_bolts/models/detection/yolo

I'm still happy to help with the code if there's interest, but I don't work for Groke Technologies anymore and I don't have the computational resources, so I cannot really continue to develop any features that we're missing to achieve a better performance on the COCO dataset.

I did go through other YOLO implementations, however, and noticed one detail that has been added at some point but I have missed: [Task Alignment Learning (TAL)](https://arxiv.org/abs/2108.07755). Assigning ground-truth boxes to anchors is based on a metric that combines both the score given to the correct class and the IoU. In case someone has the resources to continue the development, that's one thing to look at.
